entity-registry/src/main/java/com/linkedin/metadata/aspect/validation/ConditionalWriteValidator.java

```diff
@@ -0,0 +1,225 @@
+package com.linkedin.metadata.aspect.validation;
+
+import com.google.common.collect.ImmutableSet;
+import com.google.common.net.HttpHeaders;
+import com.linkedin.common.urn.Urn;
+import com.linkedin.events.metadata.ChangeType;
+import com.linkedin.metadata.aspect.AspectRetriever;
+import com.linkedin.metadata.aspect.RetrieverContext;
+import com.linkedin.metadata.aspect.SystemAspect;
+import com.linkedin.metadata.aspect.batch.BatchItem;
+import com.linkedin.metadata.aspect.batch.ChangeMCP;
+import com.linkedin.metadata.aspect.plugins.config.AspectPluginConfig;
+import com.linkedin.metadata.aspect.plugins.validation.AspectPayloadValidator;
+import com.linkedin.metadata.aspect.plugins.validation.AspectValidationException;
+import com.linkedin.metadata.aspect.plugins.validation.ValidationExceptionCollection;
+import com.linkedin.util.Pair;
+import java.time.Instant;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import javax.annotation.Nonnull;
+import lombok.Getter;
+import lombok.Setter;
+import lombok.experimental.Accessors;
+
+/**
+ * Implements conditional write semantics.
+ *
+ * <p>1. If an aspect change request includes a SystemMetadata section which includes a value for
+ * nextAspectVersion, then it must match the current SystemMetadata's value for nextAspectVersion or
+ * prevent performing the write operation.
+ *
+ * <p>2. Headers are included with the MCP for the following time-based checks * If-Modified-Since *
+ * If-Unmodified-Since
+ */
+@Setter
+@Getter
+@Accessors(chain = true)
+public class ConditionalWriteValidator extends AspectPayloadValidator {
+  public static final String DEFAULT_ASPECT_VERSION = "1";
+  public static final long DEFAULT_LAST_MODIFIED_TIME = Long.MIN_VALUE;
+  public static final String HTTP_HEADER_IF_VERSION_MATCH = "If-Version-Match";
+  public static final Set<ChangeType> CREATE_CHANGE_TYPES =
+      ImmutableSet.of(ChangeType.CREATE, ChangeType.CREATE_ENTITY);
+
+  @Nonnull private AspectPluginConfig config;
+
+  private static boolean hasTimePrecondition(ChangeMCP item) {
+    return item.getHeader(HttpHeaders.IF_MODIFIED_SINCE).isPresent()
+        || item.getHeader(HttpHeaders.IF_UNMODIFIED_SINCE).isPresent();
+  }
+
+  private static boolean hasVersionPrecondition(ChangeMCP item) {
+    return item.getHeader(HTTP_HEADER_IF_VERSION_MATCH).isPresent();
+  }
+
+  private static boolean isApplicableFilter(ChangeMCP item) {
+    if (ChangeType.RESTATE.equals(item.getChangeType())) {
+      return false;
+    }
+
+    return hasTimePrecondition(item) || hasVersionPrecondition(item);
+  }
+
+  @Override
+  protected Stream<AspectValidationException> validatePreCommitAspects(
```

Comment:
What is the transaction isolation? Since we are validating by doing a read-before-write and comparing the state, will there be a race-condition where 2 requests come in with the same nextVersion and passes this validation?

Reply:
At least for ebean supported datastores (mysql, postgres, etc) the isolation level is repeatable-read. However I believe if both transactions write to version 0, the database will throw an exception because of the write isolation which is standard. In the code, this will trigger a re-execution in a new transaction for one of the two and it would then violate the precondition and fail during the retry.